packages/sdk-react/src/components/receivables/InvoicesTable/useInvoiceRowActionMenuCell.tsx

```diff
@@ -0,0 +1,308 @@
+   * @param props.action - The action to be performed, an `InvoicesTableRowAction`.
+   */
+  onRowActionClick: (props: {
+    id: string;
+    action?: InvoicesTableRowAction;
+  }) => void;
+
+  /**
+   * The actions to be displayed in the row's action menu.
+   * If not provided, default actions will be displayed,
+   * in case if `onRowActionClick` is provided.
+   *
+   * @example
+   * ```ts
+   * {
+   *   "draft": ['view', 'edit', 'issue', 'delete'], // Optional
+   *   "issued": ['view', 'send', 'cancel', 'copyPaymentLink'], // Optional
+   *   "canceled": ['view', 'send'], // Optional
+   *   "partially_paid": ['view', 'send', 'copyPaymentLink', 'cancel'], // Optional
+   *   "overdue": ['view', 'send', 'copyPaymentLink', 'cancel', 'markUncollectible'], // Optional
+   *   "paid": ['view'], // Optional
+   *   "uncollectible": ['view'], // Optional
+   *   "expired": ['view'], // Optional
+   *   "accepted": ['view'], // Optional
+   *   "declined": ['view'], // Optional
+   *   "recurring": ['view'], // Optional
+   *   "deleted": [], // Optional
+   * }
+   * ```
+   */
+  rowActions?: Partial<InvoicesTableRowActionSchema>;
+}
+
+export const useInvoiceRowActionMenuCell = (
+  props: UseInvoiceRowActionMenuCellProps | {}
+): GridActionsColDef<ReceivableResponse> | undefined => {
+  const { data: receivableActionSchema, userIdFromAuthToken } =
+    usePermissions('receivable');
+  const { i18n } = useLingui();
+
+  if (!('onRowActionClick' in props && props.onRowActionClick)) return;
+
+  return {
+    field: 'action_menu',
+    type: 'actions',
+    headerName: t(i18n)({
+      message: 'Action menu',
+      context: 'InvoicesTableRowActionMenu',
+    }),
+    renderHeader: () => null,
+    getActions: (params) => {
+      const menuItems = getInvoiceActionMenuItems({
+        // casts, because it is not possible to resolve `type` based on multiple different enum
+        invoice: params.row as InvoiceResponsePayload,
+        actions: props.rowActions,
+        receivableActionSchema,
+        userIdFromAuthToken,
+        i18n,
+      });
+
+      return menuItems.map(({ label, action }) => (
+        <GridActionsCellItem
+          showInMenu
+          label={label}
+          onClick={(event) => {
+            event.preventDefault();
+            props.onRowActionClick({ id: params.row.id, action });
+          }}
+        />
+      ));
+    },
+  };
+};
+
+const getInvoiceActionMenuItems = ({
+  invoice,
+  actions: inActions,
+  receivableActionSchema,
+  userIdFromAuthToken,
+  i18n,
+}: {
+  invoice: InvoiceResponsePayload;
+  actions?: Partial<InvoicesTableRowActionSchema>;
+  receivableActionSchema: ActionSchema[] | undefined;
+  userIdFromAuthToken: string | undefined;
+  i18n: I18n;
+}): InvoicesTableRowActionMenuItem[] => {
+  const actions =
+    inActions?.[invoice.status] ?? DEFAULT_ACTION_LIST[invoice.status];
+
+  const allowedActions = filterInvoiceActionMenuAllowedItems(
```

Comment:
The same as the previous question. Maybe it's better to memoize the response from this function? Otherwise, we have to create a new object and make 13 loops (inside `menuItemsPermissionMap` we call `isAllowedInvoiceAction` which makes a loop for all actions and then calls `filter`)

I think memoization should work much better to save memory and avoid extra loops.

WDYT?

Reply:
The same, JS is just fast on loops. It's a kinda built-in feature. Otherwise, why would we have so many ways to iterate?

```
1. filterInvoiceActionMenuAllowedItems: 0.0029296875 ms
2. filterInvoiceActionMenuAllowedItems: 0.006103515625 ms
3. filterInvoiceActionMenuAllowedItems: 0.003173828125 ms
4. filterInvoiceActionMenuAllowedItems: 0.0029296875 ms
5. filterInvoiceActionMenuAllowedItems: 0.001953125 ms
6. filterInvoiceActionMenuAllowedItems: 0.0029296875 ms
7. filterInvoiceActionMenuAllowedItems: 0.001953125 ms
8. filterInvoiceActionMenuAllowedItems: 0.0029296875 ms
9. filterInvoiceActionMenuAllowedItems: 0.001953125 ms
10. filterInvoiceActionMenuAllowedItems: 0.0048828125 ms
11. filterInvoiceActionMenuAllowedItems: 0.001953125 ms
```